cardano-crypto-class/src/Cardano/Crypto/KES/Sum.hs

```diff
@@ -285,50 +346,50 @@ instance (KESAlgorithm d, HashAlgorithm h)
 deriving instance HashAlgorithm h => Show (VerKeyKES (SumKES h d))
 deriving instance Eq   (VerKeyKES (SumKES h d))
 
-instance (KESAlgorithm d) => NoThunks (SignKeyKES (SumKES h d))
-
-instance (KESAlgorithm d, HashAlgorithm h)
+instance (KESAlgorithm (SumKES h d), NaCl.SodiumHashAlgorithm h, SizeHash h ~ SeedSizeKES d)
       => ToCBOR (VerKeyKES (SumKES h d)) where
   toCBOR = encodeVerKeyKES
   encodedSizeExpr _size = encodedVerKeyKESSizeExpr
 
-instance (KESAlgorithm d, HashAlgorithm h)
+instance (KESAlgorithm (SumKES h d), NaCl.SodiumHashAlgorithm h, SizeHash h ~ SeedSizeKES d)
       => FromCBOR (VerKeyKES (SumKES h d)) where
   fromCBOR = decodeVerKeyKES
 
+instance (KESAlgorithm d) => NoThunks (VerKeyKES  (SumKES h d))
 
 --
 -- SignKey instances
 --
 
-deriving instance KESAlgorithm d => Show (SignKeyKES (SumKES h d))
-
-instance (KESAlgorithm d) => NoThunks (VerKeyKES  (SumKES h d))
-
-instance (KESAlgorithm d, HashAlgorithm h)
-      => ToCBOR (SignKeyKES (SumKES h d)) where
-  toCBOR = encodeSignKeyKES
-  encodedSizeExpr _size = encodedSignKeyKESSizeExpr
-
-instance (KESAlgorithm d, HashAlgorithm h)
-      => FromCBOR (SignKeyKES (SumKES h d)) where
-  fromCBOR = decodeSignKeyKES
+-- These instances would violate mlocking protections, bleeding secret keys
+-- onto the GHC heap.
```

Comment:
Someone might assume that, since mlocked memory operations are "almost morally pure", because their only effects are memory allocations / deallocations, an implementation that uses `unsafePerformIO` might be OK (similar to how `PinnedSizedBytes` and even `ByteString` provide pure APIs on top of `IO`), but it's not - even if you can make it "safe" in the sense that no access violations or memory leaks occur, in the same way that `ByteString` does, you will still leak secrets to the Haskell heap.

IMO poisoning would be the way to go.

Reply:
> you will still leak secrets to the Haskell heap.

Absolutely, there is no way to use serialization with `cborg` for mlocked memory without leaking secrets. So I totally support your desire to poison the instances. 

Only reason why I made the comment is because commented out code does not bring much value. Poisoned instances however, do.

Reply:
Sounds like we agree violently.